eta/core/image.py

```diff
@@ -105,17 +111,32 @@ def __init__(self, images=None, schema=None):
         '''
         self.images = images or []
         self.schema = schema
+        self._filename_map = {}
+        self._build_filename_map()
 
-    def __getitem__(self, idx):
-        return self.images[idx]
+    def __getitem__(self, idx_or_filename):
+        if not isinstance(idx_or_filename, numbers.Integral):
+            return self.get_image_labels_for_filename(idx_or_filename)
+        return self.images[idx_or_filename]
+
+    def __setitem__(self, idx_or_filename, image_labels):
+        if not isinstance(idx_or_filename, numbers.Integral):
+            image_labels.filename = idx_or_filename
+            if not self.has_image_labels_for_filename(idx_or_filename):
+                self.add_image_labels(image_labels)
+                return
 
-    def __setitem__(self, idx, image_labels):
         if self.has_schema:
             self._validate_image_labels(image_labels)
+
+        idx = self._to_idx(idx_or_filename)
         self.images[idx] = image_labels
+        self._filename_map[image_labels.filename] = idx
```

Comment:
This is `__setitem__`, so if some does `image_set_labels[filename] = image_labels`, then their existing labels for that filename are overwritten. This is expected.

I do acknowledge that it's possible to have multiple `ImageLabels` with the same filename in a set by doing `ImageSetLabels.add_image_labels()`. In that case, using the `__getitem__` syntax `image_set_labels[filename]` will return the last image labels with the given filename.

The filename-based access is just syntactic sugar. I didn't want to require filenames and switch to a key-value type storage structure, because there may be cases when you just want labels and don't have filenames. This was my thought at least. And, I wanted to remain backwards-compatible.